File: dest/19_interfaces.swift
// Simple Interface

protocol MyInterface {
  func bar() -> String
  /*fun foo(): Int {
    return 1
  }*/
}

class Implementation: MyInterface {
  func bar() -> String {
    return "2-Implementation"
  }
  init() {
  }
}

// Interface + Inheritance

class Parent {
  func three() -> Int {
    return 3
  }

  var four = 4
  init() {
  }
}

class Child: Parent, MyInterface {
  let five = 5
  func six() -> String {
    return "6"
  }
  func bar() -> String {
    return "2-Child"
  }
  override init() {
  }
}

// Abstract Interface + Inheritance

class AbstractParent {
  func three() -> Int {
    fatalError("Method is abstract")
  }

  var four = 4
  init() {
  }
}

class AbstractChild: Parent {
  override init() {
  }
}

class NonAbstractChild: AbstractChild, MyInterface {
  func bar() -> String {
    return "2-NonAbstractChild"
  }
  override init() {
  }
}

func main(args: [String]) {
  // Simple Interface
  let obj = Implementation()
  print(obj.foo())
  print(obj.bar())

  // Interface + Inheritance
  let child = Child()
  print(child.foo())
  print(child.bar())
  print(child.three())
  print(child.four)
  print(child.five)
  print(child.six())

  // Abstract Interface + Inheritance
  let naChild = NonAbstractChild()
  print(naChild.foo())
  print(naChild.bar())
  print(naChild.three())
  print(naChild.four)
}

main([])

File: test-src/19_interfaces.kt
// Simple Interface

interface MyInterface {
  fun bar(): String
  /*fun foo(): Int {
    return 1
  }*/
}

class Implementation : MyInterface {
  override fun bar(): String {
    return "2-Implementation"
  }
}

// Interface + Inheritance

open class Parent {
  fun three(): Int {
    return 3
  }

  var four = 4
}

class Child: Parent(), MyInterface {
  val five = 5
  fun six(): String {
    return "6"
  }
  override fun bar(): String {
    return "2-Child"
  }
}

// Abstract Interface + Inheritance

abstract class AbstractParent {
  abstract fun three(): Int

  var four = 4
}

abstract class AbstractChild: Parent() {
}

class NonAbstractChild: AbstractChild(), MyInterface {
  override fun bar(): String {
    return "2-NonAbstractChild"
  }
}

fun main(args: Array<String>) {
  // Simple Interface
  val obj = Implementation()
  println(obj.foo())
  println(obj.bar())

  // Interface + Inheritance
  val child = Child()
  println(child.foo())
  println(child.bar())
  println(child.three())
  println(child.four)
  println(child.five)
  println(child.six())

  // Abstract Interface + Inheritance
  val naChild = NonAbstractChild()
  println(naChild.foo())
  println(naChild.bar())
  println(naChild.three())
  println(naChild.four)
}
